core/src/test/scala/integration/kafka/api/ClearPurgatoryOnLeaderMovementTest.scala

```diff
@@ -0,0 +1,89 @@
+import kafka.utils.TestUtils
+import org.apache.kafka.clients.producer.ProducerRecord
+import org.apache.kafka.common.errors.NotLeaderForPartitionException
+import org.junit.Assert.{assertFalse, assertTrue}
+import org.junit.Test
+
+/**
+ * Test the case where produce requests in the purgatory should be cleared when the leader migrates to another broker.
+ */
+class ClearPurgatoryOnLeaderMovementTest extends KafkaServerTestHarness {
+
+  val topic = "topic"
+  val partition = 0
+  def generateConfigs = {
+    val overridingProps = new Properties()
+    val numServers = 2
+    overridingProps.put(KafkaConfig.NumPartitionsProp, 1.toString)
+    // Make sure the fetch request will not return.
+    overridingProps.put(KafkaConfig.ReplicaLagTimeMaxMsProp, "100000")
+    overridingProps.put(KafkaConfig.ReplicaSocketTimeoutMsProp, "100000")
+    overridingProps.put(KafkaConfig.ReplicaFetchMinBytesProp, "100000")
+    overridingProps.put(KafkaConfig.ReplicaFetchWaitMaxMsProp, "100000")
+    TestUtils.createBrokerConfigs(numServers, zkConnect, false, interBrokerSecurityProtocol = Some(securityProtocol),
+      trustStoreFile = trustStoreFile).map(KafkaConfig.fromProps(_, overridingProps))
+  }
+
+  @Test
+  def testCleanProducerPurgatoryOnBecomeFollower() {
+    TestUtils.createTopic(zkUtils, topic, 1, 2, servers, new Properties())
+    TestUtils.waitUntilLeaderIsKnown(servers, topic, partition)
+    val producer = TestUtils.createNewProducer(brokerList)
+
+    // fetch the metadata.
+    producer.partitionsFor(topic)
+    val future = producer.send(new ProducerRecord[Array[Byte], Array[Byte]](topic, "key".getBytes, "value".getBytes))
+    // Wait 10 millisends to make sure the produce request has been sent. It is ugly but we don't have a way to
+    // see if the producer has sent the produce request.
+    Thread.sleep(10)
+    assertFalse(future.isDone)
+    // Trigger a leader migration
+    if (TestUtils.isLeaderLocalOnBroker(topic, 0, servers(0)))
+      moveLeader(oldLeader = 0, newLeader = 1)
+    else
+      moveLeader(oldLeader = 1, newLeader = 0)
+    try {
+      future.get()
+      fail("Should throw ExecutionException")
+    } catch {
+      case e: ExecutionException =>
+        assertTrue(e.getCause.isInstanceOf[NotLeaderForPartitionException])
+    }
+
+    def moveLeader(oldLeader: Int, newLeader: Int) {
+      val tap = new TopicAndPartition(topic, partition)
+      new ReassignPartitionsCommand(zkUtils, Map({tap -> Seq(newLeader, oldLeader)})).reassignPartitions()
+      TestUtils.waitUntilTrue(() => zkUtils.getPartitionsBeingReassigned().isEmpty, "Failed to finish partition assgiment before timeout")
+      PreferredReplicaLeaderElectionCommand.writePreferredReplicaElectionData(zkUtils, Set(tap))
+      TestUtils.waitUntilLeaderIsElectedOrChanged(zkUtils = zkUtils, topic = topic, partition = partition, newLeaderOpt = Some(newLeader))
+      assertTrue(TestUtils.isLeaderLocalOnBroker(topic, 0, servers(newLeader)))
```

Comment:
You should use `partition` instead of `0` (they are the same, but it's confusing as the variable is used everywhere else).